File: packages/core/src/redis/redis.ts
import Redis from 'ioredis';
import { Resource } from 'sst';

import { substituteVariables } from '../utils/utils';

const redis = new Redis(
	`rediss://${Resource.RedisUser.value}:${Resource.RedisPassword.value}@${Resource.RedisHost.value}:${Resource.RedisPort.value}`,
	{
		keyPrefix: `lumi::${process.env.APP_STAGE}::`,
		lazyConnect: true,
	},
);

export class RedisHashKeys {
	private static MOMENT_SIGNED_URL = 'moment_signed_url::{momentId}';
	private static MOMENT_THUMBNAIL_SIGNED_URL = 'moment_thumbnail_signed_url::{momentId}';

	private constructor() {}

	static momentSignedUrl(momentId: string) {
		return substituteVariables(this.MOMENT_SIGNED_URL, { momentId });
	}

	static momentThumbnailSignedUrl(momentId: string) {
		return substituteVariables(this.MOMENT_THUMBNAIL_SIGNED_URL, { momentId });
	}
}
export default redis;

File: packages/go/pkg/redis/redis.go
package redis

import (
	"fmt"
	"lumi/pkg/utils"
	"os"

	"github.com/redis/go-redis/v9"
	"github.com/sst/sst/v3/sdk/golang/resource"
)

func NewRedisClient() *redis.Client {
	redisUser, err := resource.Get("RedisUser", "value")
	if err != nil {
		panic(fmt.Errorf("failed to get RedisUser: %w", err))
	}

	redisPassword, err := resource.Get("RedisPassword", "value")
	if err != nil {
		panic(fmt.Errorf("failed to get RedisPassword: %w", err))
	}

	redisHost, err := resource.Get("RedisHost", "value")
	if err != nil {
		panic(fmt.Errorf("failed to get RedisHost: %w", err))
	}

	redisPort, err := resource.Get("RedisPort", "value")
	if err != nil {
		panic(fmt.Errorf("failed to get RedisPort: %w", err))
	}

	opt, err := redis.ParseURL(fmt.Sprintf("rediss://%s:%s@%s:%s", redisUser.(string), redisPassword.(string), redisHost.(string), redisPort.(string)))
	if err != nil {
		panic(fmt.Errorf("failed to parse Redis URL: %w", err))
	}

	return redis.NewClient(opt)
}

func Key(suffix string) string {
	stage := os.Getenv("APP_STAGE")
	return fmt.Sprintf("lumi::%s::%s", stage, suffix)
}

type HashKeys struct{}

func (hk HashKeys) MomentSignedUrl(momentId string) string {
	suffix, err := utils.SubstituteVariables("moment_signed_url::{momentId}", map[string]string{"momentId": momentId})

	if err != nil {
		fmt.Println(fmt.Errorf("There was an error trying to substitute variables: %w", err))
	}

	return Key(suffix)
}

func (hk HashKeys) MomentThumbnailSignedUrl(momentId string) string {
	suffix, err := utils.SubstituteVariables("moment_thumbnail_signed_url::{momentId}", map[string]string{"momentId": momentId})

	if err != nil {
		fmt.Println(fmt.Errorf("There was an error trying to substitute variables: %w", err))
	}

	return Key(suffix)
}
